### PR TITLE
Transformer hook system for the backend

### PR DESCRIPTION
**Pull request by halgari**
This is somewhat like Event Hooks or MediatR, but different in that the result of each hook is fed into the next hook. 

The process is fairly simple, you register for one of the transformer hook, in the DI system and then the given method is called at certain parts of the backend process. We will extend this in the future as various games need hooks for parts of the apply process. For now we can insert logic into the various parts of the sorting system which should solve #299

### CODE REVIEW

File: src/NexusMods.DataModel/TransformerHooks/IBeforeMakeApplyPlan.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    /// <summary>
+    /// This method is executed after the flattening of mods into a single collection but before the
+    /// creation of the apply plan.
+    /// </summary>
+    /// <param name="files"></param>
+    /// <param name="mods"></param>
+    /// <param name="loadout"></param>
+    /// <param name="ct"></param>
+    /// <returns></returns>
+    public ValueTask<Dictionary<GamePath, (AModFile File, Mod Mod)>> BeforeMakeApplyPlan(Dictionary<GamePath, (AModFile File, Mod Mod)> files, Loadout loadout, CancellationToken ct);
```
Comment (erri120) [inline src/NexusMods.DataModel/TransformerHooks/IBeforeMakeApplyPlan.cs:21]: Inconsistent naming, should be `BeforeMakeApplyPlanAsync`.

**Review by erri120 — CHANGES_REQUESTED** [review]
The only issue I have with this design is that we need to properly document every hook. Instead of guessing which hook you'd want for your specific purpose, we should have some docs that list all hooks and explains what they do, maybe even when you should use them.

The API of these hooks could also be improved. Ideally, the methods on the interface should provide a read-only view of the data. This is already done with the `IAfterSort.AfterSortAsync` and `IBeforeSort.BeforeSortAsync` methods, which accept and return an `IEnumerable<Mod>`.

`IAfterFlatten.AfterFlattenAsync`, `IBeforeMakeApplyPlan.BeforeMakeApplyPlan.` instead provide and return a dictionary, which can be changed by the implementation at will. `GenerateFiles` actually showcases this and is a good example for the issues this API has:

- The implementation has to create a new `List<(AModFile, Mod)>` for files that get generated during this step.
- The provided dictionary gets updated while it's being iterated on, requiring the use of `.ToList()` to prevent any iteration issues, this is slow and allocates more memory than it should.
- The implementation wants to add files and has to manually call the `LoadoutManager` to add these files to the loadout.

`GenerateFiles` is supposed to "generate[s] files for mods that have a zero hash", and as such its only concern should be iterating over the existing files, finding the files with a zero hash, and ideally returning updated files.

The complexity should not be in the implementations of these hooks, but in the `TransformerRegistry`. Ideally, the APIs accept a read-only view over the data and only return the changes (added, updated, removed). The `TransformerRegistry` should use these changes to properly update the loadout.

These issues are similar to the ones encountered while working on #292, where complexity was added to the implementations instead of the manager, however I changed this to move the complexity to the manager, leaving the implementations simpler.

**Comment by halgari** [comment]
These are good points, and something I'll consider over the weekend. My plan here was to be as open-ended in the design as possible. Returning a collection allows the callee to change the input collection in any way they want. But that does make it easier to construct buggy transformers. 

Improving these interfaces is fairly simple in several cases such as the file replacement stage. But in other case like the sorting I need to think about how to craft the API. In those cases we probably want the transformers to return ISortRules instead of allowing full access to change the mod ordering however they see fit.

In addition, for performance reasons we probably want each transformer to declare their game and file filters, instead of doing a `O(transformers * files)` scan on every use.

**Comment by halgari** [comment]
Closing as this work is being done in another PR/branch

**Pull request closed without merging**